File: exercism/haskell/hexadecimal/Hexadecimal.hs
module Hexadecimal (hexToInt) where

import Data.Char (digitToInt, isHexDigit)
import Data.List (foldl')

hexToInt :: [Char] -> Int
hexToInt hex
  | any (not . isHexDigit) hex = 0
  | otherwise = foldl' (\acc digit -> 16 * acc + digitToInt digit) 0 hex

File: exercism/kotlin/hexadecimal/src/main/kotlin/Hexadecimal.kt
object Hexadecimal {
  fun toDecimal(hex: String): Int {
    return hex.fold(0, { acc, digit ->
      16 * acc + when (digit) {
        in '0' .. '9' -> (digit - '0').toInt()
        in 'a' .. 'f' -> 10 + (digit - 'a').toInt()
        in 'A' .. 'F' -> 10 + (digit - 'A').toInt()
        else -> return 0
      }
    })
  }
}
